File: Echowaves/EWBlend.m
//
//  EWBlend.m
//  Echowaves
//
//  Created by Dmitry on 1/17/14.
//  Copyright (c) 2014 Echowaves. All rights reserved.
//

#import "EWBlend.h"

@implementation EWBlend

+(void) autoCompleteFor:(NSString *)waveName
                success:(void (^)(NSArray *waveNames))success
                failure:(void (^)(NSError *error))failure {
    AFHTTPRequestOperationManager *manager = [AFHTTPRequestOperationManager manager];
    //ideally not going to need the following line, if making a request to json service
    manager.responseSerializer = [AFJSONResponseSerializer serializer];
    manager.requestSerializer = [AFJSONRequestSerializer serializer];
    
    NSDictionary *parameters = @{@"term": waveName};

    
    [manager GET:[NSString stringWithFormat:@"%@/autocomplete-wave-name.json", EWHost]
      parameters:parameters
         success:^(AFHTTPRequestOperation *operation, id responseObject) {
             success((NSArray*)responseObject);
         }
         failure:^(AFHTTPRequestOperation *operation, NSError *error) {
             NSLog(@"Error: %@", error);
             failure(error);
         }];
}

+(void) requestBlendingWith:(NSString *)waveName
                    success:(void (^)(void))success
                    failure:(void (^)(NSError *error))failure {
    AFHTTPRequestOperationManager *manager = [AFHTTPRequestOperationManager manager];
    
    // perform authentication, wave/password non blank and exist in the server side, and enter a sending loop
    
    //ideally not going to need the following line, if making a request to json service
    manager.responseSerializer = [AFJSONResponseSerializer serializer];
    manager.requestSerializer = [AFJSONRequestSerializer serializer];
    
    NSDictionary *parameters = @{@"wave_name": waveName,
                                 @"from_wave": [APP_DELEGATE waveName]};
    
    [manager POST:[NSString stringWithFormat:@"%@/request-blending.json", EWHost] parameters:parameters success:^(AFHTTPRequestOperation *operation, id responseObject) {
        success();
    } failure:^(AFHTTPRequestOperation *operation, NSError *error) {
        NSLog(@"Error: %@", error);
        NSLog(@"Response: %@", [operation.responseObject objectForKey:@"error"]);
        failure(error);        
    }];
}

+(void) confirmBlendingWith:(NSString *)waveName
                    success:(void (^)(void))success
                    failure:(void (^)(NSError *error))failure {
    AFHTTPRequestOperationManager *manager = [AFHTTPRequestOperationManager manager];
    
    // perform authentication, wave/password non blank and exist in the server side, and enter a sending loop
    
    //ideally not going to need the following line, if making a request to json service
    manager.responseSerializer = [AFJSONResponseSerializer serializer];
    manager.requestSerializer = [AFJSONRequestSerializer serializer];
    
    NSDictionary *parameters = @{@"wave_name": waveName,
                                 @"from_wave": [APP_DELEGATE waveName]};
    
    NSLog(@")))))))))))))))))wave_name:%@", waveName);
    
    [manager POST:[NSString stringWithFormat:@"%@/confirm-blending.json", EWHost]
       parameters:parameters
          success:^(AFHTTPRequestOperation *operation, id responseObject) {
              success();
          } failure:^(AFHTTPRequestOperation *operation, NSError *error) {
              NSLog(@"Error: %@", error);
              NSLog(@"Response: %@", [operation.responseObject objectForKey:@"error"]);
              failure(error);
          }];
    
}
+(void) unblendFrom:(NSString *)waveName
            success:(void (^)(void))success
            failure:(void (^)(NSError *error))failure {
    AFHTTPRequestOperationManager *manager = [AFHTTPRequestOperationManager manager];
    
    // perform authentication, wave/password non blank and exist in the server side, and enter a sending loop
    
    //ideally not going to need the following line, if making a request to json service
    manager.responseSerializer = [AFJSONResponseSerializer serializer];
    manager.requestSerializer = [AFJSONRequestSerializer serializer];
    
    NSDictionary *parameters = @{@"wave_name": waveName,
                                 @"from_wave": [APP_DELEGATE waveName]};
    
    [manager POST:[NSString stringWithFormat:@"%@/unblend.json", EWHost] parameters:parameters success:^(AFHTTPRequestOperation *operation, id responseObject) {
        success();
    } failure:^(AFHTTPRequestOperation *operation, NSError *error) {
        NSLog(@"Error: %@", error);
        NSLog(@"Response: %@", [operation.responseObject objectForKey:@"error"]);
        failure(error);
    }];
    
}

+(void) getRequestedBlends:(void (^)(NSArray *waveNames))success
                   failure:(void (^)(NSError *error))failure {
    AFHTTPRequestOperationManager *manager = [AFHTTPRequestOperationManager manager];
    //ideally not going to need the following line, if making a request to json service
    manager.responseSerializer = [AFJSONResponseSerializer serializer];
    manager.requestSerializer = [AFJSONRequestSerializer serializer];

    NSDictionary *parameters = @{@"wave_name": [APP_DELEGATE waveName]};
    
    
    [manager GET:[NSString stringWithFormat:@"%@/requested-blends.json", EWHost]
      parameters:parameters
         success:^(AFHTTPRequestOperation *operation, id responseObject) {
             success((NSArray*)responseObject);
         }
         failure:^(AFHTTPRequestOperation *operation, NSError *error) {
             NSLog(@"Error: %@", error);
             failure(error);
         }];
    
}
+(void) getUnconfirmedBlends:(void (^)(NSArray *waveNames))success
                   failure:(void (^)(NSError *error))failure {
    AFHTTPRequestOperationManager *manager = [AFHTTPRequestOperationManager manager];
    //ideally not going to need the following line, if making a request to json service
    manager.responseSerializer = [AFJSONResponseSerializer serializer];
    manager.requestSerializer = [AFJSONRequestSerializer serializer];
    
    NSDictionary *parameters = @{@"wave_name": [APP_DELEGATE waveName]};

    [manager GET:[NSString stringWithFormat:@"%@/unconfirmed-blends.json", EWHost]
      parameters:parameters
         success:^(AFHTTPRequestOperation *operation, id responseObject) {
             success((NSArray*)responseObject);
         }
         failure:^(AFHTTPRequestOperation *operation, NSError *error) {
             NSLog(@"Error: %@", error);
             failure(error);
         }];
    
}
+(void) getBlendedWith:(void (^)(NSArray *waveNames))success
                   failure:(void (^)(NSError *error))failure {
    AFHTTPRequestOperationManager *manager = [AFHTTPRequestOperationManager manager];
    //ideally not going to need the following line, if making a request to json service
    manager.responseSerializer = [AFJSONResponseSerializer serializer];
    manager.requestSerializer = [AFJSONRequestSerializer serializer];

    NSDictionary *parameters = @{@"wave_name":[APP_DELEGATE waveName]};

    
    [manager GET:[NSString stringWithFormat:@"%@/blended-with.json", EWHost]
      parameters:parameters
         success:^(AFHTTPRequestOperation *operation, id responseObject) {
             success((NSArray*)responseObject);
         }
         failure:^(AFHTTPRequestOperation *operation, NSError *error) {
             NSLog(@"Error: %@", error);
             failure(error);
         }];
    
}



@end

File: Echowaves/EWBlend.swift
//
//  EWBlend.swift
//  Echowaves
//
//  Created by D on 11/2/14.
//  Copyright (c) 2014 Echowaves. All rights reserved.
//

import Foundation

class EWBlend: EWDataModel {
    
    
    class func unblendFrom(
        waveName:String,
        currentWave: String,
        success:() -> (),
        failure:(error: NSError!) -> ()) -> () {
            let manager = AFHTTPRequestOperationManager()
            
            // perform authentication, wave/password non blank and exist in the server side, and enter a sending loop
            
            //ideally not going to need the following line, if making a request to json service
            manager.responseSerializer = AFJSONResponseSerializer() as AFJSONResponseSerializer
            manager.requestSerializer = AFJSONRequestSerializer() as AFJSONRequestSerializer
            
            
            let parameters = ["wave_name": waveName, "from_wave": currentWave]
            
            manager.POST("\(EWHost)/unblend.json", parameters:parameters,
                success: { (operation: AFHTTPRequestOperation!,responseObject: AnyObject!) in
                    success();
                },
                failure: { (operation: AFHTTPRequestOperation!,error: NSError!) in
                    println("Error: \(error.localizedDescription)")
                    failure(error: error)
            })
            
    }
    
    
    class func getBlendedWith(
        success:(waveNames:NSArray) -> (),
        failure:(error: NSError!) -> ()) -> () {
            
            let manager = AFHTTPRequestOperationManager()
            //ideally not going to need the following line, if making a request to json service
            manager.responseSerializer = AFJSONResponseSerializer() as AFJSONResponseSerializer
            manager.requestSerializer = AFJSONRequestSerializer() as AFJSONRequestSerializer
            
            let parameters = ["wave_name":APP_DELEGATE.currentWaveName]
            
            
            manager.GET("\(EWHost)/blended-with.json",
                parameters: parameters,
                success: { (operation: AFHTTPRequestOperation!,responseObject: AnyObject!) in
                    success(waveNames: responseObject as NSArray)
                },
                failure: { (operation: AFHTTPRequestOperation!,error: NSError!) in
                    println("Error: \(error.localizedDescription)")
                    failure(error: error)
            })
    }
    
    
    class func acceptBlending(
        origMyWaveName: String,
        myWaveName: String,
        friendWaveName:String,
        success:() -> (),
        failure:(error: NSError!) -> ()) -> () {
            
            let manager = AFHTTPRequestOperationManager()
            
            // perform authentication, wave/password non blank and exist in the server side, and enter a sending loop
            
            //ideally not going to need the following line, if making a request to json service
            manager.responseSerializer = AFJSONResponseSerializer() as AFJSONResponseSerializer
            manager.requestSerializer = AFJSONRequestSerializer() as AFJSONRequestSerializer
            
            let parameters =
            ["orig_my_wave_name": origMyWaveName,
                "my_wave_name": myWaveName,
                "friend_wave_name": friendWaveName]
            
            manager.POST("\(EWHost)/accept_blending.json",
                parameters:parameters,
                success: { (operation: AFHTTPRequestOperation!,responseObject: AnyObject!) in
                    success();
                },
                failure: { (operation: AFHTTPRequestOperation!,error: NSError!) in
                    println("Error: \(error.localizedDescription)")
                    failure(error: error)
            })
    }
}
